desc/coils.py

```diff
@@ -560,6 +636,192 @@ def from_symmetry(cls, coils, NFP, sym=False):
+                    coords[:, -1][0],
+                    tempx,
+                    tempy,
+                    tempz,
+                    method=method,
+                    name=names[i],
+                )
+            )
+
+        return CoilSet(*coils, NFP=NFP)
+
+    def save_in_MAKEGRID_format(self, coilsFilename, NFP=None, grid=None):
+        """Save CoilSet of as a MAKEGRID-formatted coil txtfile.
+
+        By default, each coil is assigned to the same Coilgroup in MAKEGRID
+        with the name "Modular". For more details see the MAKEGRID documentation
+        https://princetonuniversity.github.io/STELLOPT/MAKEGRID.html
+
+        Note: if a nested CoilSet, will flatten it first before saving
+        #TODO: name each group based off of CoilSet name?
+        #TODO: have CoilGroup be automatically assigned based off of
+        CoilSet if current coilset is a collection of coilsets?
+
+        Parameters
+        ----------
+        filename : str or path-like
+            path save CoilSet as a file in MAKEGRID txt format
+        NFP : int, default None
+            If > 1, assumes that the CoilSet is the coils for a coilset
+            with a nominal discrete toroidal symmetry of NFP, and will
+            put that NFP in the periods line of the coils file generated.
+            defaults to self.NFP
+        grid: Grid, ndarray, int,
+            Grid of sample points along each coil to save.
+            if None, will default to each coils self._grid
+        """
+        NFP = self.NFP if NFP is None else NFP
+        coils = flatten_list(self.coils)  # flatten any nested coilsets
+        assert (
+            int(len(coils) / NFP) == len(coils) / NFP
+        ), "Number of coils in coilset must be evenly divisible by NFP!"
+
+        header = (
+            # number of field period
+            "periods "
+            + str(NFP)
+            + "\n"
+            + "begin filament\n"
+            # not 100% sure of what this line is, neither is MAKEGRID,
+            # but it is needed and expected by other codes
+            # "The third line is read by MAKEGRID but ignored"
+            # https://princetonuniversity.github.io/STELLOPT/MAKEGRID.html
+            + "mirror NIL"
+        )
+        footer = "end\n"
+
+        x_arr = []
+        y_arr = []
+        z_arr = []
+        currents_arr = []
+        coil_end_inds = []  # indices where the coils end, need to track these
+        # to place the coilgroup number and name later, which MAKEGRID expects
+        # at the end of each individual coil
+        for i in range(int(len(coils))):
+            coil = coils[i]
+            coords = coil.compute("x", basis="xyz", grid=grid)["x"]
+
+            contour_X = np.asarray(coords[0:-1, 0])
+            contour_Y = np.asarray(coords[0:-1, 1])
+            contour_Z = np.asarray(coords[0:-1, 2])
+
+            currents = np.ones_like(contour_X) * float(coil.current)
+            # close the curves
```

Comment:
thanks for catching that. Technically I think it was ok because when I index coords I did `[:-1]` so the last point was not included, but it only works assuming the grid had endpoint=True, I fixed it now